File: estimation/ekf.py
import numpy as np


def wrap_angle(angle: float) -> float:
    return (angle + np.pi) % (2 * np.pi) - np.pi


def motion_model(x: np.ndarray, u: np.ndarray, dt: float) -> np.ndarray:
    px, py, theta = x
    v, omega = u
    return np.array([
        px + v * np.cos(theta) * dt,
        py + v * np.sin(theta) * dt,
        wrap_angle(theta + omega * dt),
    ])


def jacobian_F(x: np.ndarray, u: np.ndarray, dt: float) -> np.ndarray:
    _, _, theta = x
    v, _ = u
    return np.array([
        [1.0, 0.0, -v * np.sin(theta) * dt],
        [0.0, 1.0,  v * np.cos(theta) * dt],
        [0.0, 0.0,  1.0],
    ])


def measurement_model(x: np.ndarray) -> np.ndarray:
    return np.array([x[0], x[1]])


def jacobian_H() -> np.ndarray:
    return np.array([
        [1.0, 0.0, 0.0],
        [0.0, 1.0, 0.0],
    ])


def ekf_predict(x: np.ndarray, P: np.ndarray, u: np.ndarray, Q: np.ndarray, dt: float):
    x_pred = motion_model(x, u, dt)
    F = jacobian_F(x, u, dt)
    P_pred = F @ P @ F.T + Q
    return x_pred, P_pred


def ekf_update(x_pred: np.ndarray, P_pred: np.ndarray, z: np.ndarray, R: np.ndarray):
    H = jacobian_H()
    z_pred = measurement_model(x_pred)

    y = z - z_pred
    S = H @ P_pred @ H.T + R
    K = P_pred @ H.T @ np.linalg.inv(S)

    x_upd = x_pred + K @ y
    x_upd[2] = wrap_angle(x_upd[2])

    I = np.eye(len(x_pred))
    P_upd = (I - K @ H) @ P_pred

    return x_upd, P_upd

File: web/src/estimation/ekf.js
function wrapAngle(angle) {
  const twoPi = 2 * Math.PI;
  return ((((angle + Math.PI) % twoPi) + twoPi) % twoPi) - Math.PI;
}

export function motionModel(state, control, dt) {
  const [x, y, theta] = state;
  const [v, omega] = control;

  return [
    x + v * Math.cos(theta) * dt,
    y + v * Math.sin(theta) * dt,
    wrapAngle(theta + omega * dt),
  ];
}

function jacobianF(state, control, dt) {
  const [, , theta] = state;
  const [v] = control;

  return [
    [1, 0, -v * Math.sin(theta) * dt],
    [0, 1, v * Math.cos(theta) * dt],
    [0, 0, 1],
  ];
}

function jacobianH() {
  return [
    [1, 0, 0],
    [0, 1, 0],
  ];
}

function measurementModel(state) {
  return [state[0], state[1]];
}

function matMul(a, b) {
  const rows = a.length;
  const cols = b[0].length;
  const inner = b.length;
  const out = Array.from({ length: rows }, () => Array(cols).fill(0));

  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      for (let k = 0; k < inner; k++) {
        out[r][c] += a[r][k] * b[k][c];
      }
    }
  }

  return out;
}

function matAdd(a, b) {
  return a.map((row, r) => row.map((value, c) => value + b[r][c]));
}

function matSub(a, b) {
  return a.map((row, r) => row.map((value, c) => value - b[r][c]));
}

function transpose(m) {
  return m[0].map((_, c) => m.map((row) => row[c]));
}

function matVecMul(m, v) {
  return m.map((row) => row.reduce((sum, value, i) => sum + value * v[i], 0));
}

function vecAdd(a, b) {
  return a.map((value, i) => value + b[i]);
}

function vecSub(a, b) {
  return a.map((value, i) => value - b[i]);
}

function identity(n) {
  return Array.from({ length: n }, (_, r) =>
    Array.from({ length: n }, (_, c) => (r === c ? 1 : 0))
  );
}

function inverse2x2(m) {
  const [[a, b], [c, d]] = m;
  const det = a * d - b * c;

  if (Math.abs(det) < 1e-9) {
    throw new Error("EKF update failed: singular innovation matrix");
  }

  const invDet = 1 / det;
  return [
    [d * invDet, -b * invDet],
    [-c * invDet, a * invDet],
  ];
}

export function ekfPredict(state, covariance, control, processNoise, dt) {
  const predictedState = motionModel(state, control, dt);
  const F = jacobianF(state, control, dt);
  const predictedCovariance = matAdd(
    matMul(matMul(F, covariance), transpose(F)),
    processNoise
  );

  return { state: predictedState, covariance: predictedCovariance };
}

export function ekfUpdate(predictedState, predictedCovariance, measurement, measurementNoise) {
  const H = jacobianH();
  const predictedMeasurement = measurementModel(predictedState);
  const innovation = vecSub(measurement, predictedMeasurement);
  const S = matAdd(
    matMul(matMul(H, predictedCovariance), transpose(H)),
    measurementNoise
  );
  const K = matMul(
    matMul(predictedCovariance, transpose(H)),
    inverse2x2(S)
  );

  const updatedState = vecAdd(predictedState, matVecMul(K, innovation));
  updatedState[2] = wrapAngle(updatedState[2]);

  const I = identity(predictedState.length);
  const updatedCovariance = matMul(matSub(I, matMul(K, H)), predictedCovariance);

  return { state: updatedState, covariance: updatedCovariance };
}

export function diag3(a, b, c) {
  return [
    [a, 0, 0],
    [0, b, 0],
    [0, 0, c],
  ];
}
